### ISSUE TITLE
#outboundproxy All instances have full, unlogged, unconfigurable outbound internet access

### ISSUE THREAD

**Issue by tinnightcap** [nubisproject/nubis-meta#30]
All instances have full, unlogged, unconfigurable outbound internet access

Currently all instances are granted the ability to initiate connections to any external internet website they wish. These connections are not logged and if a type of outbound connection was identified as one used by threat actors to exfiltrate date, it could not be blocked.

Typically when a threat actor succeeds at exploiting a weakness in a system, and they wish to exfiltrate data (user's stored passwords, system passwords, etc) from that compromised machine they will instruct the instance to push the data that they want out to a location on the internet. By allowing all instances the ability to initiate connections to all locations on the internet, a threat actor will have an easy time exfiltrating data from any compromised host.

Additionally, since outbound access is wide open, these potential data exfiltrations can not be logged to detect potential security incidents.

Finally, since there is nothing governing these outbound connections, in the case of a security incident where an attacker is exfiltrating data, we would have no ability to stop the exfiltration. We also have no ability to pre-emptively block outbound access to known threat actor sites.

Recommendations:

    Introduce http/https proxy allowing specific outbound patterns.

The steps to implement this security control are:

    Create an http/https proxy server (for example squid)
    Grant that proxy server outbound access to the internet
    For a given product or service, task the product developers with enumerating the outbound sites their product initiates connections to
    Render this site list or site list pattern into rules for the proxy server
    Configure the product or service to use the http/https proxy server instead of initiating connections itself
    Change the security group associated with the product or service from one which allows outbound internet access to one which allows outbound connections only to the proxy server.

This process of creating product specific rules and cutting products over to using the proxy service can be done in a decoupled fashion from other products. Each product team can create their own rules (as they are in the best position to know what outbound connections the product makes).

For products which have not yet enumerated the sites they connect to, they can be pointed to the proxy server with a blanket "allow all" rule while they develop their rules. This doesn't help in preventing data exfiltration but it is useful both in building the rules as well as logging potential data exfiltration.

By introducing an http/https proxy server we also gain the ability to log failed outbound connection attempts, thereby detecting compromised machines.

An http/https proxy server also gives us a powerful tool to surgically stop outbound traffic during a security incident, allowing us to potentially keep the product or service up and available while at the same time preventing exfiltration of data on the servers that makeup that product/service.

**Comment by tinnightcap** [comment]
TL;DR, This will take me approximately one week to implement for all VPCs in all accounts. If this is a top priority (drop everything) I should be able to have it implemented by the end of next week (Sep 24th 2015) as long as I have the decision by COB Thursday (tomorrow).

The Full Story:
As it is configured today[1] in the datacenters, all Bugzilla nodes are
allowed connections through the proxies to anywhere on allowed ports[2].
This translates to, all current outbound connections for Bugzilla nodes
are logged on allowed ports. All other connections are blocked[3].
It is important to note that the only thing that the proxies are doing
for Bugzilla currently is to log outbound connections on the list of
approved ports[2].

To replicate this functionality in Nubis I would need to
  a) Create puppet module for nubis base (no good module on puppetforge,
would basically copy current module from datacenter puppet and modify to
suit)
  b) Build AMI for proxys utilizing a new nubis-proxy GitHub repository
  c) Use existing consul service discovery to expose proxy service to
hosts as a DNS name
  d) Use existing Fluent collectors to ship logs off to the existing S3
log collection buckets
  e) Use consul key/value store to load/update configuration on running
proxies (technically not necessary as current datacenters allows 'all'.
AKA this is additional, potentially unnecessary, plan-b work)
  f) Leverage existing DataDog integration for monitoring (zero effort
required)

This would be much faster than plumbing additional VPNs and attempting
to integrate with datacenter puppetmasters. After a collaboration with
Gozer, we determined that it is non-trivial to attempt to integrate
ephemeral instances with puppet in a reliable way. Additionally the
semi-reliable method we discussed would require more engineering time
than the above proposed method. I can go into great detail on this point if required.

Props to limed for his assistance helping me to understand the current
datacenter proxy deployment.

1. http_access allow src_(Bugzilla) dst_(Bugzilla)
   src (Bugzilla CIDRs) (ie: 10.8.70.0/24, 10.8.82.0/24)
   dst all
2. acl safe_ports port 21 80 443 873 8000 9418
   acl ssl_ports port 443 873 9418
3. http_access deny all (last line in config)

**Comment by tinnightcap** [comment]
This work as far as the current scope is concerned has been completed. Alternate issues have been filed for any additional tasks remaining.
